urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/Compression/LZMA/Encoder.java

```diff
@@ -0,0 +1,1726 @@
+                        + posState]);
+        int repMatchPrice = matchPrice
+                + SevenZip.Compression.RangeCoder.Encoder
+                        .GetPrice1(_isRep[_state]);
+        
+        if (matchByte == currentByte)
+        {
+            final int shortRepPrice = repMatchPrice
+                    + GetRepLen1Price(_state, posState);
+            if (shortRepPrice < _optimum[1].Price)
+            {
+                _optimum[1].Price = shortRepPrice;
+                _optimum[1].MakeAsShortRep();
+            }
+        }
+        
+        int lenEnd = lenMain >= repLens[repMaxIndex] ? lenMain
+                : repLens[repMaxIndex];
+        
+        if (lenEnd < 2)
+        {
+            backRes = _optimum[1].BackPrev;
+            return 1;
+        }
+        
+        _optimum[1].PosPrev = 0;
+        
+        _optimum[0].Backs0 = reps[0];
+        _optimum[0].Backs1 = reps[1];
+        _optimum[0].Backs2 = reps[2];
+        _optimum[0].Backs3 = reps[3];
+        
+        int len = lenEnd;
+        do
+        {
+            _optimum[len--].Price = kIfinityPrice;
+        } while (len >= 2);
+        
+        for (i = 0; i < Base.kNumRepDistances; i++)
+        {
+            int repLen = repLens[i];
+            if (repLen < 2)
+            {
+                continue;
+            }
+            final int price = repMatchPrice
+                    + GetPureRepPrice(i, _state, posState);
+            do
+            {
+                final int curAndLenPrice = price
+                        + _repMatchLenEncoder.GetPrice(repLen - 2, posState);
+                final Optimal optimum = _optimum[repLen];
+                if (curAndLenPrice < optimum.Price)
+                {
+                    optimum.Price = curAndLenPrice;
+                    optimum.PosPrev = 0;
+                    optimum.BackPrev = i;
+                    optimum.Prev1IsChar = false;
+                }
+            } while (--repLen >= 2);
+        }
+        
+        int normalMatchPrice = matchPrice
+                + SevenZip.Compression.RangeCoder.Encoder
+                        .GetPrice0(_isRep[_state]);
+        
+        len = repLens[0] >= 2 ? repLens[0] + 1 : 2;
+        if (len <= lenMain)
+        {
+            int offs = 0;
+            while (len > _matchDistances[offs])
+            {
+                offs += 2;
+            }
+            for (;; len++)
+            {
+                final int distance = _matchDistances[offs + 1];
+                final int curAndLenPrice = normalMatchPrice
+                        + GetPosLenPrice(distance, len, posState);
+                final Optimal optimum = _optimum[len];
+                if (curAndLenPrice < optimum.Price)
+                {
+                    optimum.Price = curAndLenPrice;
+                    optimum.PosPrev = 0;
+                    optimum.BackPrev = distance + Base.kNumRepDistances;
+                    optimum.Prev1IsChar = false;
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 807:813, 1182:1188.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.